### PR TITLE
add beats_translate_move ControlEncoder

### PR DESCRIPTION
**Pull request by mxmilkiib**
See https://github.com/mixxxdj/mixxx/issues/12337

### CODE REVIEW

**Review by ronso0 — CHANGES_REQUESTED** [review]
Thanks for working on this, adn congratulations for your first PR!

For every code change to be included in Mixxx we need contributor's permissions.
> Please sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) and comment here when you have done so. It gives us permission to distribute your contribution under the GPL v2 or later license and the Apple Mac App Store. It is also helpful for us to have contact information for contributors in case we may need it in the future.

I've left some comments / suggestions.

File: src/engine/controls/bpmcontrol.cpp
Comment (mxmilkiib) [reply src/engine/controls/bpmcontrol.cpp:197]: Good to know, thanks. I think I added the prefix to move beyond a red squiggly line in VSCodium, but I guess some other difference created that.

File: src/engine/controls/bpmcontrol.cpp
Comment (mxmilkiib) [reply src/engine/controls/bpmcontrol.cpp:203]: Aha, that's why it went one way only, I guess.

File: src/engine/controls/bpmcontrol.cpp
Comment (mxmilkiib) [reply src/engine/controls/bpmcontrol.cpp:216]: Nice compactification, I'll switch to that

File: src/skin/legacy/tooltips.cpp
Comment (mxmilkiib) [reply src/skin/legacy/tooltips.cpp:415]: 🫡

**Comment by ronso0** [comment]
About code reviews: please keep them open, it's up to the reviewer to check and mark as resolved (besides no one will see your reply if you resolve a conversation ;)

**Comment by ronso0** [comment]
Please also check the CI runners (at the bottom of this page) once they are finished.

Worth reading to avoid cleanup work later on:
https://github.com/mixxxdj/mixxx/wiki/Contribution-Guidelines
and this one about automatic ode checking (pre-commit)
https://github.com/mixxxdj/mixxx/wiki/Using%20Git#set-up-automatic-code-checking

Thanks for signing the CLA!
You'll find your name in the contributor list in the About dialog in the next release.

File: src/engine/controls/bpmcontrol.cpp
Comment (ronso0) [reply src/engine/controls/bpmcontrol.cpp]: Oh, superfluous `)` at the end.

**Comment by mxmilkiib** [comment]
Thanks for all the feedback!

On the current build fail due to code style, I'm confused;

image

https://github.com/mixxxdj/mixxx/actions/runs/7051273531/job/19194088455?pr=12376#step:6:133

Should I change the code to the format in green? But that would make it different from the rest of the formatting.

https://github.com/mxmilkiib/mixxx/blob/beats_translate_move/src/engine/controls/bpmcontrol.cpp#L90

**Comment by ronso0** [comment]
I understand this desire, though we try to avoid needless mass-formatting. Usually, after being merged, it creates conflicts in other PRs.

**Comment by ronso0** [comment]
Same for bpmcontrol.cpp I'd say, break after every comma.

**Comment by mxmilkiib** [comment]
Does the code style linter need updating then?

**Comment by ronso0** [comment]
Actually it's configured well, but every now and then it has a bad day or something and its suggestions are... weird.
We don't obey it 100% but it makes life much easier.

**Comment by ronso0** [comment]
You can also ignore certain checks locally (once, for the current command) if they keep you from commiting, for example disable `clang-format` if it insists on the long line after you broke it in pieces:
`SKIP=clang-format git [commit|push|whatever]`

**Comment by daschuer** [comment]
> Does the code style linter need updating then?

We apply clang-format twice, to respect user line breaks. First run: Unbreak lines with breaks not following the rules. Second Run: break the break all lines the exceeds 100 at <80. 
https://github.com/mixxxdj/mixxx/blob/0f0deb18160547bcd96e0a60d5fc7c7a9d4ba306/.clang-format#L7

**Comment by ronso0** [comment]
yeah, but it failed here. Its proposal is 126 chars + indentation.

File: src/engine/controls/bpmcontrol.cpp
Comment (ronso0) [inline src/engine/controls/bpmcontrol.cpp:201]: I think we need to cast to integer here so the behaviour still matches the description. I.e. round fractional values so we get a constant step size.
```suggestion
    v = std::round(v);
    if (v == 0) {
```

**Review by ronso0 — CHANGES_REQUESTED** [review]

**Comment by ronso0** [comment]
One more comment.
And the linebreaks please, then I'll take a final look.

**Comment by mxmilkiib** [comment]
How could I make the line
```
        const mixxx::audio::FrameDiff_t frameOffset = sampleOffset / mixxx::kEngineChannelCount;
```
not be 98 characters wide?

Assuming that's the main offender.

File: src/engine/controls/bpmcontrol.cpp
Comment (ronso0) [inline src/engine/controls/bpmcontrol.cpp:91]: I think this will make clang-format happy
```suggestion
    m_pTranslateBeatsMove =
            new ControlEncoder(ConfigKey(group, "beats_translate_move"), false);
```

**Comment by ronso0** [comment]
> I'm not sure why I shouldn't?

IME we only do that to get code that was merged since opeing a PR / starting a branch, especially for long-standing PRs, in order to get other fixes and/or fix merge conflicts.
You can do that whenever you like, I was just wondering what you expect from it.

The line-break issue still needs to be resolved, one `connect` arg per line. If you can't resist just change the calls above and below, as well, I don't mind. Not many (and only trivial) merge conflicts to be expected with this PR.

File: src/engine/controls/bpmcontrol.h
Comment (ronso0) [reply src/engine/controls/bpmcontrol.h:145]: @daschuer could that be a consequence of PCH (which I have disabled locally)?

**Comment by ronso0** [comment]
Okay, I'm sorry to hear that.

I'll adopt this PR then.
Thank you for getting this started and for your work!

**Comment by ronso0** [comment]
Please don't delete your branch, I'll push to it and someone else needs to do the final review and merge.

**Comment by ronso0** [comment]
@mixxxdj/developers May someone please take a look?
I adopted this and added another commit, therefore I won't merge myself. Thanks!

File: src/engine/controls/bpmcontrol.cpp
Comment (JoergAtGithub) [inline src/engine/controls/bpmcontrol.cpp:194]:
```suggestion
    if (v <= 0) {
        return;
    }
    slotTranslateBeatsMove(-1.0);
```
Otherwise the code is identical to `slotTranslateBeatsLater` and the behavior of the function would be different, than before.

**Review by JoergAtGithub — CHANGES_REQUESTED** [review]

File: src/engine/controls/bpmcontrol.cpp
Comment (ronso0) [reply src/engine/controls/bpmcontrol.cpp]: :facepalm: 
Thank you. I pushed the fix:
later: +1
earlier: -1

File: src/engine/controls/bpmcontrol.cpp
Comment (ronso0) [reply src/engine/controls/bpmcontrol.cpp:130]: Yep.

**Comment by ywwg** [comment]
looks good, thank you @mxmilkiib and @ronso0 !

**Comment by ywwg** [comment]
@ronso0 and @JoergAtGithub are your notes addressed?

**Pull request merged**
